ccl/streamingccl/streamproducer: TestStreamPartition failed

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9799281?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9799281?buildTab=artifacts#/) on release-23.1 @ [ec0b5087392d49980a2141315aae62bb2563348f](https://github.com/cockroachdb/cockroach/commits/ec0b5087392d49980a2141315aae62bb2563348f):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 58890 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamPartition
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ec6a3f034611d027670bad7acfd7b3dd/logTestStreamPartition337479638
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestStreamPartition/stream-table-cursor-error
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #88429 ccl/streamingccl/streamproducer: TestStreamPartition failed [C-test-failure O-robot T-cdc branch-release-22.1]
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27382

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9819927?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9819927?buildTab=artifacts#/) on release-23.1 @ [0d8f02757d44e05435eb8de80a311fcfe625ef1b](https://github.com/cockroachdb/cockroach/commits/0d8f02757d44e05435eb8de80a311fcfe625ef1b):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 60946 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamPartition
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ec6a3f034611d027670bad7acfd7b3dd/logTestStreamPartition489391464
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestStreamPartition/stream-table-cursor-error
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #88429 ccl/streamingccl/streamproducer: TestStreamPartition failed [C-test-failure O-robot T-cdc branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9823870?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9823870?buildTab=artifacts#/) on release-23.1 @ [86a86b15c13afe0a580deed35ce7dafea7adad78](https://github.com/cockroachdb/cockroach/commits/86a86b15c13afe0a580deed35ce7dafea7adad78):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 60650 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamPartition
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ec6a3f034611d027670bad7acfd7b3dd/logTestStreamPartition2455180566
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestStreamPartition/stream-table-cursor-error
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9890597?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9890597?buildTab=artifacts#/) on release-23.1 @ [fa8d237846052ae72cc44b3eaebe1f38c89834f8](https://github.com/cockroachdb/cockroach/commits/fa8d237846052ae72cc44b3eaebe1f38c89834f8):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 57967 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamPartition
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ec6a3f034611d027670bad7acfd7b3dd/logTestStreamPartition2388411531
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestStreamPartition/stream-table-cursor-error
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #102520 ccl/streamingccl/streamproducer: TestStreamPartition failed [C-test-failure O-robot T-disaster-recovery branch-release-23.1.0]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamproducer.TestStreamPartition [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9965449?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9965449?buildTab=artifacts#/) on release-23.1 @ [4226a83871bbce776bc9389fca5cf084b4bb7632](https://github.com/cockroachdb/cockroach/commits/4226a83871bbce776bc9389fca5cf084b4bb7632):


```
=== RUN   TestStreamPartition
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/9ce24496530642208197dda818d65c9e/logTestStreamPartition2047022242
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestStreamPartition
    replication_stream_test.go:478: -- test log scope end --
    replication_stream_test.go:478: Leaked goroutine: goroutine 4772422 [IO wait]:
        internal/poll.runtime_pollWait(0x7f25d2c85e28, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc011148d80?, 0xc0076a5200?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc011148d80, {0xc0076a5200, 0x10, 0x10})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc011148d80, {0xc0076a5200?, 0xc01244de48?, 0x4a7a92?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc012adbc18, {0xc0076a5200?, 0xc0159f8d20?, 0xc0159f8d20?})
        	GOROOT/src/net/net.go:183 +0x45
        github.com/jackc/pgconn.(*PgConn).CancelRequest(0xc00a862580, {0x70c0c50?, 0xc0159f8d20})
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:934 +0x3db
        github.com/jackc/pgconn.(*PgConn).asyncClose.func1()
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:667 +0x10f
        created by github.com/jackc/pgconn.(*PgConn).asyncClose
        	github.com/jackc/pgconn/external/com_github_jackc_pgconn/pgconn.go:658 +0x65
        Leaked goroutine: goroutine 4772457 [select]:
        github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1()
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:47 +0x6c
        created by github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch
        	github.com/jackc/pgconn/internal/ctxwatch/external/com_github_jackc_pgconn/internal/ctxwatch/context_watcher.go:46 +0x12c
--- FAIL: TestStreamPartition (22.33s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #102520 ccl/streamingccl/streamproducer: TestStreamPartition failed [C-test-failure O-robot T-disaster-recovery branch-release-23.1.0]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamPartition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/streamingccl/replicationtestutils/replication_helpers.go
+++ b/pkg/ccl/streamingccl/replicationtestutils/replication_helpers.go
@@ -141,11 +141,11 @@ func (rf *ReplicationFeed) consumeUntil(
 			for {
 				msg, haveMoreRows := rf.f.Next()
 				if !haveMoreRows {
-					if rf.f.Error() != nil {
-						if errPred(rf.f.Error()) {
+					if err := rf.f.Error(); err != nil {
+						if errPred(err) {
 							return nil
 						}
-						return rf.f.Error()
+						return err
 					}
 					return errors.Newf("ran out of rows after processing %d rows", rowCount)
 				}

--- a/pkg/ccl/streamingccl/streamproducer/event_stream.go
+++ b/pkg/ccl/streamingccl/streamproducer/event_stream.go
@@ -81,8 +81,10 @@ func (s *eventStream) Start(ctx context.Context, txn *kv.Txn) error {
 
 	s.acc = s.mon.MakeBoundAccount()
 
-	// errCh consumed by ValueGenerator and is signaled when go routines encounter error.
-	s.errCh = make(chan error)
+	// errCh is buffered to ensure the sender can send an error to
+	// the buffer, without waiting, when the channel receiver is not waiting on
+	// the channel.
+	s.errCh = make(chan error, 1)
 
 	// Events channel gets RangeFeedEvents and is consumed by ValueGenerator.
 	s.eventsCh = make(chan kvcoord.RangeFeedMessage)
@@ -159,6 +161,9 @@ func (s *eventStream) Start(ctx context.Context, txn *kv.Txn) error {
 }
 
 func (s *eventStream) maybeSetError(err error) {
+	// Only send the error if the channel is empty, else it's ok to swallow the
+	// error because the first error in the channel will shut down the event
+	// stream.
 	select {
 	case s.errCh <- err:
 	default:
@@ -223,7 +228,6 @@ func (s *eventStream) Close(ctx context.Context) {
 		// Note: error in close is normal; we expect to be terminated with context canceled.
 		log.Errorf(ctx, "partition stream %d terminated with error %v", s.streamID, err)
 	}
-
 	s.sp.Finish()
 }
 

--- a/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go
+++ b/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go
@@ -133,6 +133,7 @@ func (d *partitionStreamDecoder) decode() {
 func (f *pgConnReplicationFeedSource) Next() (streamingccl.Event, bool) {
 	f.mu.Lock()
 	defer f.mu.Unlock()
+	// First check if there exists a decoded event.
 	if e := f.mu.codec.pop(); e != nil {
 		return e, true
 	}
@@ -342,9 +343,7 @@ func TestStreamPartition(t *testing.T) {
 	srcTenant.SQL.Exec(t, `
 CREATE DATABASE d;
 CREATE TABLE d.t1(i int primary key, a string, b string);
-CREATE TABLE d.t2(i int primary key, a string, b string);
 INSERT INTO d.t1 (i) VALUES (42);
-INSERT INTO d.t2 (i) VALUES (42);
 USE d;
 `)
 
@@ -355,18 +354,22 @@ USE d;
 
 	const streamPartitionQuery = `SELECT * FROM crdb_internal.stream_partition($1, $2)`
 	t1Descr := desctestutils.TestingGetPublicTableDescriptor(h.SysServer.DB(), srcTenant.Codec, "d", "t1")
-	t2Descr := desctestutils.TestingGetPublicTableDescriptor(h.SysServer.DB(), srcTenant.Codec, "d", "t2")
 
 	t.Run("stream-table-cursor-error", func(t *testing.T) {
-		skip.WithIssue(t, 102286)
+
+		srcTenant.SQL.Exec(t, `
+CREATE TABLE d.t2(i int primary key, a string, b string);
+INSERT INTO d.t2 (i) VALUES (42);
+`)
 		_, feed := startReplication(ctx, t, h, makePartitionStreamDecoder,
 			streamPartitionQuery, streamID, encodeSpec(t, h, srcTenant, initialScanTimestamp, hlc.Timestamp{}, "t2"))
 		defer feed.Close(ctx)
+		t2Descr := desctestutils.TestingGetPublicTableDescriptor(h.SysServer.DB(), srcTenant.Codec, "d", "t2")
+		expected := replicationtestutils.EncodeKV(t, srcTenant.Codec, t2Descr, 42)
+		feed.ObserveKey(ctx, expected.Key)
 
 		subscribedSpan := h.TableSpan(srcTenant.Codec, "t2")
 		// Send a ClearRange to trigger rows cursor to return internal error from rangefeed.
-		// Choose 't2' so that it doesn't trigger error on other registered span in rangefeeds,
-		// affecting other tests.
 		_, err := kv.SendWrapped(ctx, h.SysServer.DB().NonTransactionalSender(), &kvpb.ClearRangeRequest{
 			RequestHeader: kvpb.RequestHeader{
 				Key:    subscribedSpan.Key,
@@ -375,13 +378,8 @@ USE d;
 		})
 		require.Nil(t, err)
 
-		expected := replicationtestutils.EncodeKV(t, srcTenant.Codec, t2Descr, 42)
-		feed.ObserveKey(ctx, expected.Key)
 		feed.ObserveError(ctx, func(err error) bool {
-			return strings.Contains(err.Error(), "unexpected MVCC history mutation") ||
-				// TODO(casper): disabled this once we figured out why we have context cancellation
-				// emitted from ingestion side.
-				strings.Contains(err.Error(), "context canceled")
+			return strings.Contains(err.Error(), "unexpected MVCC history mutation")
 		})
 	})
 

